gapis/api/vulkan/transform_wait_for_perfetto.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package vulkan implementes the API interface for the Vulkan graphics library.
+
+package vulkan
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/google/gapid/core/event/task"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapir"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/transform2"
+	"github.com/google/gapid/gapis/replay"
+	"github.com/google/gapid/gapis/replay/builder"
+	"github.com/google/gapid/gapis/service"
+	"github.com/google/gapid/gapis/trace"
+)
+
+var _ transform2.Transform = &waitForPerfetto{}
+
+type waitForPerfetto struct {
+	traceOptions  *service.TraceOptions
+	signalHandler *replay.SignalHandler
+	buffer        *bytes.Buffer
+}
+
+type FenceCallback = func(ctx context.Context, request *gapir.FenceReadyRequest)
+
+func newWaitForPerfetto(traceOptions *service.TraceOptions, signalHandler *replay.SignalHandler, buffer *bytes.Buffer) *waitForPerfetto {
+	return &waitForPerfetto{
+		traceOptions:  traceOptions,
+		signalHandler: signalHandler,
+		buffer:        buffer,
+	}
+}
+
+func (perfettoTransform *waitForPerfetto) RequiresAccurateState() bool {
+	return false
+}
+
+func (perfettoTransform *waitForPerfetto) BeginTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	return inputCommands, nil
+}
+
+func (perfettoTransform *waitForPerfetto) ClearTransformResources(ctx context.Context) {
+	// Do nothing
+}
+
+func (perfettoTransform *waitForPerfetto) EndTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	inputCommands = append(inputCommands, createVkDeviceWaitIdleCommandsForDevices(ctx, inputState)...)
+
+	// Melih TODO: We have to explain this magic id
```

Comment:
It's probably the CmdNoID after the masking.

Reply:
This is not related at all to the cmdID. This is just a uint32 that is being stuffed into a CmdID, then taken back out. (should just change the parameter type, since it is never used as a command id).

This is a fence ID.
```
fenceID := uint32(id)
		b.Wait(fenceID)
		tcb := func(p *gapir.FenceReadyRequest) {
			callback(ctx, p)
		}
		return b.RegisterFenceReadyRequestCallback(fenceID, tcb)
```
This is literally just a magic number that maps the Fence to the callback for when the fence is hit.